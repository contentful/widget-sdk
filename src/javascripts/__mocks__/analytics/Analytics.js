export const track = jest.fn();
export const enable = jest.fn();
