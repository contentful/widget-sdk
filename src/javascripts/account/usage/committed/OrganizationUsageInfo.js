import React from 'react';
import PropTypes from 'prop-types';
import { TextLink } from '@contentful/forma-36-react-components';

import { shorten } from 'utils/NumberUtils.es6';

export default class OrganizationUsageInfo extends React.Component {
  static propTypes = {
    totalUsage: PropTypes.number.isRequired,
    includedLimit: PropTypes.number.isRequired
  };

  render() {
    const { totalUsage, includedLimit } = this.props;
    return (
      <div>
        <h2>Total number of API requests</h2>
        <div className="usage-page__total-usage">{totalUsage.toLocaleString('en-US')}</div>
        <div className="usage-page__limit">
          <span className="usage-page__included-limit">{`${shorten(includedLimit)} included`}</span>
          {totalUsage > includedLimit && (
            <span className="usage-page__overage">
              {` + ${(totalUsage - includedLimit).toLocaleString('en-US')} overage`}
            </span>
          )}
          <TextLink
            href="https://www.contentful.com/r/knowledgebase/fair-use/"
            target="_blank"
            className="usage-page__learn-more-link">
            Learn more
          </TextLink>
        </div>
      </div>
    );
  }
}
