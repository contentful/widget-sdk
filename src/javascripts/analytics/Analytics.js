import * as Config from 'Config';
import * as Snowplow from 'analytics/snowplow';
import stringifySafe from 'json-stringify-safe';
import { prepareUserData } from 'analytics/UserData';
import _ from 'lodash';
import segment from 'analytics/segment';
import { eventExists, transformEvent } from 'analytics/transform';
import * as logger from 'services/logger';
import * as analyticsConsole from 'analytics/analyticsConsole';
import * as random from '../utils/Random';
import { clearSequenceContext, initSequenceContext } from './sequenceContext';

function removeCircularRefs(obj) {
  return JSON.parse(stringifySafe(obj));
}

/**
 * @ngdoc service
 * @name analytics
 * @description
 * This service exposes an API for event tracking.
 *
 * Call to `enable` enables tracking and initializes
 * session data with user's details. There are three
 * tracking methods that are also used to collect
 * session data (`trackContextChange`, `trackStateChange`)
 * . Session data can be obtained with `getSessionData`.
 *
 * The rest of tracking is realised with calls to
 * `track` method.
 *
 * Calling `enable` doesn't mean that the events
 * will be sent to Segment and Snowplow automatically. We perform
 * an environment check to determine if we should do
 * networking.
 *
 * Once disabled, this service cannot be enabled
 * again.
 */

const ANALYTICS_ENVS = ['production', 'staging', 'preview'];
const VALUE_UNKNOWN = {};

let env = Config.env;
const session = {};
let isEnabled = false;

// Ugly but it's super tricky to simulate environment.
// Better ideas needed.
export const __testOnlySetEnv = (_env) => {
  env = _env;
};

/**
 * @ngdoc method
 * @name analytics#enable
 * @description
 * Starts event tracking
 */
export const enable = _.once((user, segmentLoadOptions) => {
  if (isEnabled) {
    return;
  }

  isEnabled = true;

  if (ANALYTICS_ENVS.includes(env)) {
    segment.enable(segmentLoadOptions);
    Snowplow.enable();
  }

  identify(prepareUserData(removeCircularRefs(user)));
  track('global:app_loaded');
});

/**
 * @ngdoc method
 * @name analytics#getSessionData
 * @param {string|array?} path
 * @param {any?} defaultValue
 * @returns {object}
 * @description
 * Gets session data. If `path` is provided then
 * extract specific nested value.
 */
export function getSessionData(path, defaultValue) {
  return _.get(session, path || [], defaultValue);
}

/**
 * @ngdoc method
 * @name analytics#track
 * @param {string} event
 * @param {object?} data
 * @description
 * Sends tracking event (with optionally provided data) to Segment and Snowplow
 * if it is on the valid events list.
 */

export function track(event, data) {
  if (!isEnabled) {
    return;
  }

  if (!eventExists(event)) {
    track('tracking:invalid_event', { event });
    return;
  }

  try {
    data = _.isObject(data) ? _.cloneDeep(data) : {};
    data = removeCircularRefs(Object.assign({}, getBasicPayload(), data));

    const transformedData = transformEvent(event, data);

    segment.track(event, transformedData);
    Snowplow.track(event, transformedData);
    analyticsConsole.add(event, transformedData);
    logEventPayloadSize(event, transformedData);
  } catch (error) {
    // ensure no errors caused by analytics will break business logic
    logger.logError('Unexpected error during event tracking', {
      error,
      message: error.message,
      event,
      data,
    });
  }
}

/**
 * This method helps to identify misuse of the analytics module
 * when SDK Value Objects sent in the event payload instead of
 * Data Transfer Objects. That results in the performance degradations
 * of the web app caused by heavy payload serialization in wootric (loaded in segment)
 *
 */
function logEventPayloadSize(event, safePayload) {
  if (typeof window.requestIdleCallback !== 'undefined') {
    window.requestIdleCallback(() => {
      try {
        const size = JSON.stringify(safePayload).length;

        // any of the payload fields has methods on the first level
        const hasMethods = Object.entries(safePayload || {})
          .flatMap(([_, v]) => Object.values(v || {}))
          .some((v) => _.isFunction(v));

        if (size > 5000 || hasMethods) {
          logger.logWarn('Potentially bloated tracking event payload', {
            event,
            size,
            hasMethods,
          });
        }
      } catch (error) {
        // ignore error
      }
    });
  }
}

/**
 * @description This function allows you to extend user's details
 * which will be sent to segment (and automatically to Intercom)
 * @param {object} params - object with new data
 */
export const updateUserInSegment = identify;

/**
 * @ngdoc method
 * @name analytics#identify
 * @param {object} extension
 * @description
 * Sets or extends session user data. Identifying
 * data is also set on Segment's client.
 */
function identify(extension) {
  session.user = session.user || {};
  const rawUserData = _.merge(session.user, extension || {});

  // we set up organization immediately, if it is not set up yet.
  // the reason for it – we might not have a space, but belong to an
  // organization. This happens, for example, after the user signs up
  // and when they have no spaces – so adding it to the session
  // automatically enriches all events with an organizationId.
  if (!session.organization) {
    // we've already removed all circular references
    // we default to the first org the user belongs to as the "current org".
    // this behaviour is the same as what the sidepanel does when it doesn't find a "current org"
    session.organization = _.get(session, 'user.organizationMemberships[0].organization');
  }

  // We need to remove the list of organization memberships as this array gets
  // flattened when it is passed to Intercom and creates a lot of noise
  const user = _.omitBy(rawUserData, (val) => _.isArray(val) || _.isObject(val));

  const userId = getSessionData('user.sys.id');

  if (userId && user) {
    segment.identify(userId, user);
    Snowplow.identify(userId);
  }

  sendSessionDataToConsole();
}

/**
 * @ngdoc method
 * @name analytics#trackContextChange
 * @param {object} space
 * @param {object} organization
 * @description
 * Sets or replaces session space and organization
 * data. Pass `null` when leaving context.
 *
 * `null` must be explicitly passed to unset the current
 * space/org contexts.
 */

export function trackContextChange(space, organization) {
  if (space) {
    session.space = removeCircularRefs(space);
  } else if (space === null) {
    session.space = null;
  }

  if (organization) {
    session.organization = removeCircularRefs(organization);
  } else if (organization === null) {
    session.organization = null;
  }

  sendSessionDataToConsole();
  track(space ? 'global:space_changed' : 'global:space_left');
}

/**
 * @ngdoc method
 * @name analytics#trackStateChange
 * @param {object} state
 * @param {object} params
 * @param {object} from
 * @param {object} fromParams
 * @description
 * Sets or replaces session navigation data.
 * Accepts arguments of `$stateChangeSuccess`
 * handler. Current state is set as a page on
 * the Segment's client.
 */
export function trackStateChange(state, params, from, fromParams) {
  if (!isEnabled) {
    return;
  }

  const data = (session.navigation = removeCircularRefs({
    state: state.name,
    params,
    fromState: from ? from.name : null,
    fromStateParams: fromParams || null,
  }));

  sendSessionDataToConsole();

  if (state.name === 'spaces.detail.entries.list') {
    initSequenceContext({ sequence_key: random.id() });
  } else {
    clearSequenceContext();
  }

  track('global:state_changed', data);
  segment.page(state.name, params);
}

function getBasicPayload() {
  // IMPORTANT: Do not add anything here without ensuring that it won't end up in any Snowplow
  //  events which might be considered invalid with additional properties unknown to their schema!
  return _.pickBy(
    {
      userId: getSessionData('user.sys.id', VALUE_UNKNOWN),
      spaceId: getSessionData('space.sys.id', VALUE_UNKNOWN),
      organizationId: getSessionData('organization.sys.id', VALUE_UNKNOWN),
      currentState: getSessionData('navigation.state', VALUE_UNKNOWN),
    },
    (val) => val !== VALUE_UNKNOWN
  );
}

function sendSessionDataToConsole() {
  analyticsConsole.setSessionData(session);
}
