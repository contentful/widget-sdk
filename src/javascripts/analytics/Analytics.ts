import * as Config from 'Config';
import segment, { Plan, transformSnowplowToSegmentData } from 'analytics/segment';
import * as snowplow from 'analytics/snowplow';
import stringifySafe from 'json-stringify-safe';
import { prepareUserData } from 'analytics/UserData';
import _ from 'lodash';
import {
  eventExists,
  getSnowplowSchemaForEvent,
  transformEventForSnowplow,
} from 'analytics/transform';
import { TransformedEventData, EventData } from './types';
import { captureError, captureWarning } from 'core/monitoring';
import * as analyticsConsole from 'analytics/analyticsConsoleController';
import * as random from '../utils/Random';
import { clearSequenceContext, initSequenceContext } from './sequenceContext';

const requestIdleCallback = (window as any).requestIdleCallback || _.noop;

function removeCircularRefs(obj): Record<string, unknown> {
  return JSON.parse(stringifySafe(obj));
}

/**
 * @ngdoc service
 * @name analytics
 * @description
 * This service exposes an API for event tracking.
 *
 * Call to `enable` enables tracking and initializes
 * session data with user's details. There are three
 * tracking methods that are also used to collect
 * session data (`trackContextChange`, `trackStateChange`)
 * . Session data can be obtained with `getSessionData`.
 *
 * The rest of tracking is realised with calls to
 * `track` method.
 *
 * Calling `enable` doesn't mean that the events
 * will be sent to Segment and Snowplow automatically. We perform
 * an environment check to determine if we should do
 * networking.
 *
 * Once disabled, this service cannot be enabled
 * again.
 */

const ANALYTICS_ENVS = ['production', 'staging', 'preview'];
const VALUE_UNKNOWN = {};

let env = Config.env;
const session: Record<string, unknown> = {};
let isEnabled = false;

// Ugly but it's super tricky to simulate environment.
// Better ideas needed.
export const __testOnlySetEnv = (_env) => {
  env = _env;
};

/**
 * @ngdoc method
 * @name analytics#enable
 * @description
 * Starts event tracking
 */
export const enable = _.once((user, segmentLoadOptions) => {
  if (isEnabled) {
    return;
  }

  isEnabled = true;

  if (ANALYTICS_ENVS.includes(env)) {
    segment.enable(segmentLoadOptions);
    snowplow.enable();
  }

  identify(prepareUserData(removeCircularRefs(user)));
  track('global:app_loaded');
});

/**
 * Gets specific nested session data at `path`.
 */
export function getSessionData(path: string, defaultValue?: unknown): unknown {
  return _.get(session, path, defaultValue);
}

/**
 * @deprecated Use `Analytics.tracking.…` methods instead which takes the final event data without
 *  sending it through the deprecated event transformers.
 *
 * Sends tracking event with provided data to Segment and Snowplow if registered properly for the event.
 */

export function track(event: string, data?: EventData): void {
  if (!isEnabled) {
    return;
  }

  if (!eventExists(event)) {
    track('tracking:invalid_event', { event });
    return;
  }

  try {
    data = _.isObject(data) ? _.cloneDeep(data) : {};
    data = removeCircularRefs(Object.assign({}, getBasicPayload(), data));

    const transformedSnowplowData = transformEventForSnowplow(event, data);
    const transformedSegmentData = transformSnowplowToSegmentData(
      event,
      transformedSnowplowData,
      getEnvironmentId()
    );

    segment.track(event, transformedSegmentData);
    snowplow.track(event, transformedSnowplowData);

    analyticsConsole.add(event, { rawData: data, transformedSegmentData, transformedSnowplowData });

    logEventPayloadSize(event, transformedSnowplowData);
  } catch (error) {
    // ensure no errors caused by analytics will break business logic
    captureError(error, {
      extra: {
        event,
        data,
      },
    });
  }
}

/**
 * Exposes all Segment typewriter tracking functions for each Segment plan (schema).
 * E.g. `Analytics.tracking.editorLoaded(props);`
 *
 * Use this instead of `Analytics.track()`. Update event registration and remove
 * transformation from transform.ts when migrating an event from `Analytics.track() and
 * use `migratedLegacyEvent()` if the event should still be tracked to Snowplow too.
 *
 * TODO: Can we enrich `props` with default props `organization_key`, `space_key` and `environment_key`?
 *  Obstacles to this approach:
 *  1. Some events that are org rather than space scoped might only have a `organization_key`
 *  2. Legacy events from Snowplow -> Segment migration have other default properties,
 *    `executing_user_id`, `space_id` and `organization_id`
 *  3. Consider all props being wrapped in `props.data` on at least most legacy events, so default props
 *     had to be attached to either `props.data` if it exists or `props` otherwise.
 *  4. Legacy Snowplow migration events use camelCaseKeys like `executingUserId` in their Segment schema
 *     This should be cleaned up soon as convention for all Segment schemas is also snake_case.
 */
export const tracking: Plan = _.mapValues(segment.plan, (planFn, planKey) => (props) => {
  try {
    // Track to Segment by using original plan function:
    planFn(props);

    // Depending on the `data` format we get and the event we're dealing with, we've got to ensure it's
    // in the right format for Snowplow where we never wrapped { data } while this is done in most
    // migrated events' Snowplow schemas due to an old tracking bug where data was accidentally wrapped.
    const likeTransformedData: TransformedEventData = _.isObject(props.data)
      ? props
      : { data: props };
    const snowplowSchema = getSnowplowSchemaForEvent(planKey);

    if (snowplowSchema) {
      snowplow.track(planKey, likeTransformedData);
    }

    // TODO: Catch errors via `onViolation` Segment TypeWriter option and display in analytics console.
    analyticsConsole.add(planKey, { rawData: props, transformedSnowplowData: likeTransformedData });
  } catch (error) {
    // ensure no errors caused by analytics will break business logic
    captureError(error, {
      extra: {
        planKey,
        data: props,
      },
    });
  }
});

/**
 * Returns props that should be used by all new Segment events tracked via typewriter.
 *
 * TODO: Add these automatically inside `Analytics.typewriter.` functions so that tracking
 *  calls don't have to call this function.
 */
export function defaultEventProps() {
  return _.pickBy(
    {
      space_key: getSessionData('space.sys.id', VALUE_UNKNOWN),
      organization_key: getSessionData('organization.sys.id', VALUE_UNKNOWN),
      environment_key: getSessionData('environment.sys.id', VALUE_UNKNOWN),
    },
    (val) => val !== VALUE_UNKNOWN
  );
}

/**
 * This method helps to identify misuse of the analytics module
 * when SDK Value Objects sent in the event payload instead of
 * Data Transfer Objects. That results in the performance degradations
 * of the web app caused by heavy payload serialization in wootric (loaded in segment)
 *
 */
function logEventPayloadSize(eventName: string, safePayload: TransformedEventData) {
  requestIdleCallback(() => {
    try {
      const { contexts: contextEvents = [], ...primaryPayload } = safePayload;
      const primaryEventSize = JSON.stringify(primaryPayload).length;
      // Context events are Snowplow specific
      const contextEventsSize = JSON.stringify(contextEvents).length - 2; // -2 to account for `[]`

      // any of the payload fields has methods on the first level
      const hasMethods = Object.entries(safePayload || {})
        .flatMap(([_, v]) => Object.values(v || {}))
        .some((v) => _.isFunction(v));

      if (primaryEventSize > 5000 || contextEventsSize > 15000 || hasMethods) {
        captureWarning(new Error('Potentially bloated tracking event payload'), {
          extra: {
            event: eventName,
            primaryEventSize,
            contextEventsSize,
            contextEventsCount: contextEvents.length,
            hasMethods,
          },
        });
      }
    } catch (error) {
      // ignore error
    }
  });
}

/**
 * This function allows you to extend user's details which will
 * be sent to segment (and automatically to Intercom)
 */
export const updateUserInSegment = identify;

/**
 * Sets or extends session user data. Identifying data is also set on Segment's client.
 */
function identify(extension?: object): void {
  session.user = session.user || {};
  const rawUserData = _.merge(session.user, extension || {});

  // we set up organization immediately, if it is not set up yet.
  // the reason for it – we might not have a space, but belong to an
  // organization. This happens, for example, after the user signs up
  // and when they have no spaces – so adding it to the session
  // automatically enriches all events with an organizationId.
  if (!session.organization) {
    // we've already removed all circular references
    // we default to the first org the user belongs to as the "current org".
    // this behaviour is the same as what the sidepanel does when it doesn't find a "current org"
    session.organization = _.get(session, 'user.organizationMemberships[0].organization');
  }

  // We need to remove the list of organization memberships as this array gets
  // flattened when it is passed to Intercom and creates a lot of noise
  const user = _.omitBy(rawUserData, (val) => _.isArray(val) || _.isObject(val));

  const userId = getSessionData('user.sys.id') as string;
  const userSignature = getSessionData('user.intercomUserSignature');

  if (userId && user) {
    segment.identify(userId, user, {
      integrations: {
        Intercom: { user_hash: userSignature }, // for identity verification purpose
      },
    });
    snowplow.identify(userId);
  }

  sendSessionDataToConsole();
}

// TODO: Get rid of `session` data usage and narrow down bare minimum types, ideally pass only IDs.
interface MaybeSys extends Record<string, any> {
  id: string;
}
interface MaybeEntity extends Record<string, any> {
  sys: MaybeSys;
}
type TrackingContext = Partial<
  Record<'organization' | 'space' | 'environment', MaybeEntity | null>
>;

/**
 * Sets or replaces session space and organization data. Pass `null` when leaving context.
 * `null` must be explicitly passed to unset the current space/organization/environment contexts.
 */
export function trackContextChange({ organization, space, environment }: TrackingContext): void {
  if (environment) {
    session.environment = removeCircularRefs(environment);
  } else if (environment === null) {
    session.environment = null;
  }

  if (space) {
    session.space = removeCircularRefs(space);
  } else if (space === null) {
    session.space = null;
  }

  if (organization) {
    session.organization = removeCircularRefs(organization);
  } else if (organization === null) {
    session.organization = null;
  }

  sendSessionDataToConsole();
  // TODO: This might cause unnecessary events if space didn't really change!
  track(space ? 'global:space_changed' : 'global:space_left');
}

// TODO: Either combine trackStateChange() state and params or just pass a string for state.name
type NavigationSate = {
  name: string;
};

/**
 * Sets or replaces session navigation data.
 * Accepts arguments of `$stateChangeSuccess`
 * handler. Current state is set as a page on
 * the Segment's client.
 */
export function trackStateChange(
  state: NavigationSate,
  params: object,
  from: NavigationSate,
  fromParams: object
): void {
  if (!isEnabled) {
    return;
  }

  const data = (session.navigation = removeCircularRefs({
    state: state.name,
    params,
    fromState: from ? from.name : null,
    fromStateParams: fromParams || null,
  }));

  sendSessionDataToConsole();

  // TODO: Get rid of this logic and generate a `sequence_key` close to the views where we do the tracking.
  if (['spaces.detail.entries.list', 'spaces.environment.entries.list'].includes(state.name)) {
    initSequenceContext({ sequence_key: random.id() });
  } else {
    clearSequenceContext();
  }

  track('global:state_changed', data);
  segment.page(state.name, params);
}

type BasicEventData = {
  userId?: string;
  spaceId?: string;
  organizationId?: string;
  currentState?: string;
};

function getBasicPayload(): BasicEventData {
  // IMPORTANT: Do not add anything here without ensuring that it won't end up in any Snowplow
  //  events which might be considered invalid with additional properties unknown to their schema!
  return _.pickBy(
    {
      userId: getSessionData('user.sys.id', VALUE_UNKNOWN),
      spaceId: getSessionData('space.sys.id', VALUE_UNKNOWN),
      organizationId: getSessionData('organization.sys.id', VALUE_UNKNOWN),
      currentState: getSessionData('navigation.state', VALUE_UNKNOWN),
    },
    (val) => val !== VALUE_UNKNOWN
  );
}

function getEnvironmentId() {
  return getSessionData('environment.sys.id', undefined) as string | undefined;
}

function sendSessionDataToConsole() {
  analyticsConsole.setSessionData(session);
}
