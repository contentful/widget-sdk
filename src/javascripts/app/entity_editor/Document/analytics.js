import * as Analytics from 'analytics/Analytics';
import changedEntityFieldPaths from './changedEntityFieldPaths';
import { cmaGetEntity } from './api';
import * as logger from 'services/logger';
export async function trackEditConflict({
  spaceEndpoint,
  localEntity,
  localEntityFetchedAt,
  changedLocalEntity,
}) {
  let remoteEntity;
  try {
    const { sys } = localEntity;
    remoteEntity = await cmaGetEntity(spaceEndpoint, sys.type, sys.id);
  } catch (e) {
    if (e.code === 'NotFound') {
      // TODO: track with affordable analytics
      return;
    } else {
      // @ts-ignore
      logger.logError('Could not fetch remote entity for CmaDocument edit conflict tracking');
      return;
    }
  }
  const data = conflictEntitiesToTrackingData();
  Analytics.track('entity_editor:edit_conflict', data);
  function conflictEntitiesToTrackingData() {
    return {
      entityId: localEntity.sys.id,
      entityType: localEntity.sys.type,
      localChangesFieldPaths: changedEntityFieldPaths(
        localEntity.fields,
        changedLocalEntity.fields
      ).map((path) => path.join(':')),
      remoteChangesSinceLocalEntityFieldPaths: changedEntityFieldPaths(
        localEntity.fields,
        remoteEntity.fields || {}
      ).map((path) => path.join(':')),
      localEntityVersion: localEntity.sys.version,
      remoteEntityVersion: remoteEntity.sys.version,
      localEntityUpdatedAtTstamp: localEntity.sys.updatedAt,
      remoteEntityUpdatedAtTstamp: remoteEntity.sys.updatedAt,
      remoteEntityUpdatedByUserId: remoteEntity.sys.updatedBy.sys.id,
      localEntityLastFetchedAtTstamp: localEntityFetchedAt.toISOString(),
      isConflictAutoResolvable: false,
      autoConflictResolutionVersion: 1,
    };
  }
}
