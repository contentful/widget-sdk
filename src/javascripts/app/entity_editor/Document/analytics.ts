import { Entity } from './types';
import * as Analytics from 'analytics/Analytics';
import changedEntityFieldPaths from './changedEntityFieldPaths';
import { EntityRepo } from 'data/CMA/EntityRepo';
import * as logger from 'services/logger';

export async function trackEditConflict({
  entityRepo,
  localEntity,
  localEntityFetchedAt,
  changedLocalEntity,
  ...options
}: {
  entityRepo: EntityRepo;
  localEntity: Entity;
  localEntityFetchedAt: Date;
  changedLocalEntity: Entity;
  remoteEntity?: Entity;
}) {
  let remoteEntity: Entity | null;
  try {
    remoteEntity =
      options.remoteEntity || (await entityRepo.get(localEntity.sys.type, localEntity.sys.id));
  } catch (e) {
    if (e.code === 'NotFound') {
      // TODO: Track deleted entity conflict with affordable analytics. move to separate function
      //  To avoid accidentally tracking `BadRequest` errors as a conflict.
      return;
    } else {
      // @ts-ignore
      logger.logError('Could not fetch remote entity for CmaDocument edit conflict tracking', e);
      return;
    }
  }

  // Do not track if there's no conflict. E.g. if we call this function due to a `BadRequest`
  // error to accommodate for possible entity deletion kind of conflict.
  if (localEntity.sys.version !== remoteEntity.sys.version) {
    const data = conflictEntitiesToTrackingData();
    Analytics.track('entity_editor:edit_conflict', data);
  }

  function conflictEntitiesToTrackingData() {
    return {
      entityId: localEntity.sys.id,
      entityType: localEntity.sys.type,
      localChangesFieldPaths: changedEntityFieldPaths(
        localEntity.fields,
        changedLocalEntity.fields
      ).map((path) => path.join(':')),
      remoteChangesSinceLocalEntityFieldPaths: changedEntityFieldPaths(
        localEntity.fields,
        remoteEntity.fields || {}
      ).map((path) => path.join(':')),
      localEntityVersion: localEntity.sys.version,
      remoteEntityVersion: remoteEntity.sys.version,
      localEntityUpdatedAtTstamp: localEntity.sys.updatedAt,
      remoteEntityUpdatedAtTstamp: remoteEntity.sys.updatedAt,
      remoteEntityUpdatedByUserId: remoteEntity.sys.updatedBy.sys.id,
      localEntityLastFetchedAtTstamp: localEntityFetchedAt.toISOString(),
      isConflictAutoResolvable: false,
      autoConflictResolutionVersion: 1,
    };
  }
}
