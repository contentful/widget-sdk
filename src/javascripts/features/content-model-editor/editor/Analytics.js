import * as Analytics from 'analytics/Analytics';
import { get } from 'lodash';

export const trackAddedField = (contentType, field) => {
  Analytics.track('content_modelling:field_added', {
    contentTypeId: contentType.sys.id,
    contentTypeName: contentType.name,
    fieldId: field.id,
    fieldName: field.name,
    fieldType: field.type,
    fieldItemType: get(field, 'items.type') || null,
    fieldLocalized: field.localized,
    fieldRequired: field.required,
  });
};

export const trackEnforcedButtonClick = (err) => {
  // If we get reason(s), that means an enforcement is present
  const reason = get(err, 'body.details.reasons', null);

  Analytics.track('entity_button:click', {
    entityType: 'contentType',
    enforced: Boolean(reason),
    reason,
  });
};
