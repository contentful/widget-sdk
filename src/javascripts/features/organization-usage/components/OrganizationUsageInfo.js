import React from 'react';
import {
  Typography,
  Paragraph,
  TextLink,
  Heading,
  SkeletonContainer,
  SkeletonDisplayText,
  SkeletonBodyText,
} from '@contentful/forma-36-react-components';
import tokens from '@contentful/forma-36-tokens';
import { track } from 'analytics/Analytics';
import { css } from 'emotion';
import { shorten } from 'utils/NumberUtils';
import { useUsageState } from '../hooks/usageContext';

const styles = {
  heading: css({
    color: '#536171',
    fontWeight: tokens.fontWeightNormal,
  }),
  usageNumber: css({
    color: '#253545',
    fontSize: tokens.fontSize3Xl,
    fontWeight: tokens.fontWeightMedium,
  }),
  overageNumber: css({
    color: '#9F6312',
    fontSize: tokens.fontSizeS,
  }),
};

export const OrganizationUsageInfo = () => {
  const { totalUsage, apiRequestIncludedLimit, isLoading } = useUsageState();

  const limitedUsage = !!apiRequestIncludedLimit;

  const handleClick = () => {
    track('usage:fair_use_policy_clicked');
  };

  return (
    <Typography>
      <Heading element="h2" className={styles.heading}>
        Total API requests
      </Heading>
      {isLoading && (
        <SkeletonContainer>
          <SkeletonDisplayText numberOfLines={1} />
          <SkeletonBodyText numberOfLines={4} offsetTop={55} />
        </SkeletonContainer>
      )}
      {!isLoading && (
        <>
          <Paragraph data-test-id="org-usage-total" className={styles.usageNumber}>
            {totalUsage.toLocaleString('en-US')}
            {limitedUsage && totalUsage > apiRequestIncludedLimit && (
              <small data-test-id="org-usage-overage" className={styles.overageNumber}>
                {` +${(totalUsage - apiRequestIncludedLimit).toLocaleString('en-US')} overage`}
              </small>
            )}
          </Paragraph>
          <Paragraph>
            {limitedUsage ? (
              <>
                {'Total API calls made this month from a '}
                <strong data-test-id="org-usage-limit">{shorten(apiRequestIncludedLimit)}</strong>
                {
                  '/month quota. This number includes CMA, CDA, CPA, and GraphQL requests. The use of Contentful is subject to our '
                }
              </>
            ) : (
              'This number includes CMA, CDA, CPA, and GraphQL requests. The use of Contentful is subject to our '
            )}
            <TextLink
              href="https://www.contentful.com/r/knowledgebase/fair-use/"
              target="_blank"
              data-test-id="fair_use_policy_link"
              onClick={handleClick}
              rel="noopener noreferrer">
              Fair Use Policy
            </TextLink>
            .
          </Paragraph>
        </>
      )}
    </Typography>
  );
};
