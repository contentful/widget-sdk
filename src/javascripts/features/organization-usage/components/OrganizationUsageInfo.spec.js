import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { track } from 'analytics/Analytics';
import { OrganizationUsageInfo } from './OrganizationUsageInfo';

describe('OrganisationUsageInfoNew', () => {
  const defaultProps = {
    totalUsage: 23000,
    includedLimit: 2000000,
  };

  const renderComp = (props) => {
    return render(<OrganizationUsageInfo {...props} />);
  };

  it('should render', () => {
    const { container } = renderComp(defaultProps);
    expect(container).toMatchSnapshot();
  });

  it('should render correct data', () => {
    const { getByTestId } = renderComp(defaultProps);
    const total = getByTestId('org-usage-total');
    const limit = getByTestId('org-usage-limit');
    expect(total.textContent).toBe('23,000');
    expect(limit.textContent).toBe('2M');
  });

  it('should render overage', () => {
    const overageProps = {
      totalUsage: 2500000,
      includedLimit: 2000000,
    };

    const { getByTestId } = renderComp(overageProps);
    const overage = getByTestId('org-usage-overage');
    expect(overage).toHaveTextContent('+500,000 overage');
  });

  it('tracks fair use policy click', () => {
    const { getByTestId } = renderComp(defaultProps);
    fireEvent.click(getByTestId('fair_use_policy_link'));
    expect(track).toHaveBeenCalledTimes(1);
    expect(track).toHaveBeenCalledWith('usage:fair_use_policy_clicked');
  });
});
