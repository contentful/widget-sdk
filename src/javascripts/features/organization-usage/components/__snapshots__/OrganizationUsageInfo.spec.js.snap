// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganisationUsageInfo should render 1`] = `
.emotion-0 {
  color: #536171;
  font-weight: 400;
}

.emotion-1 {
  color: #253545;
  font-size: 2.1875rem;
  font-weight: 600;
}

<div>
  <div
    class="Typography__Typography___1ZUfE"
    data-test-id="cf-ui-text-container"
  >
    <h2
      class="Heading__Heading___83D3K emotion-0 f36-margin-bottom--m"
      data-test-id="cf-ui-heading"
    >
      Total API requests
    </h2>
    <p
      class="Paragraph__Paragraph___2aO-9 emotion-1 f36-margin-bottom--m"
      data-test-id="org-usage-total"
    >
      23,000
    </p>
    <p
      class="Paragraph__Paragraph___2aO-9 f36-margin-bottom--m"
      data-test-id="cf-ui-paragraph"
    >
      Total API calls made this month from a 
      <strong
        data-test-id="org-usage-limit"
      >
        2M
      </strong>
      /month quota. This number includes CMA, CDA, CPA, and GraphQL requests. To learn about utility limits, read the 
      <a
        class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
        data-test-id="fair_use_policy_link"
        href="https://www.contentful.com/r/knowledgebase/fair-use/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          class="TabFocusTrap__TabFocusTrap___39Vty"
          tabindex="-1"
        >
          Fair Use Policy
        </span>
      </a>
      .
    </p>
  </div>
</div>
`;
